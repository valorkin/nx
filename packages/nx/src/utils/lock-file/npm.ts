import { LockFileData, PackageDependency } from './lock-file-type';
import { sortObject, hashString } from './utils';

type PackageMeta = {
  path: string;
  optional?: boolean;
  dev?: boolean;
};

type Dependencies = Record<string, Omit<PackageDependency, 'packageMeta'>>;

type NpmDependency = {
  version: string;
  resolved: string;
  integrity: string;
  requires?: Record<string, string>;
  dependencies?: Record<string, NpmDependency>;
  dev?: boolean;
  devOptional?: boolean;
  optional?: boolean;
};

type NpmLockFile = {
  name?: string;
  lockfileVersion: number;
  requires?: boolean;
  packages: Dependencies;
  dependencies?: Record<string, NpmDependency>;
};

/**
 * Parses package-lock.json file to `LockFileData` object
 *
 * @param lockFile
 * @returns
 */
export function parseNpmLockFile(lockFile: string): LockFileData {
  const { packages, dependencies, ...metadata } = JSON.parse(lockFile);
  return {
    dependencies: mapPackages(packages),
    lockFileMetadata: {
      metadata,
      rootPackage: packages[''],
    },
    hash: hashString(lockFile),
  };
}

// Maps /node_modules/@abc/def with version 1.2.3 => @abc/def > @abc/dev@1.2.3
function mapPackages(packages: Dependencies): LockFileData['dependencies'] {
  const mappedPackages: LockFileData['dependencies'] = {};

  Object.entries(packages).forEach(([key, { dev, optional, ...value }]) => {
    // skip root package
    if (!key) {
      return;
    }
    // key can have several instances of 'node_modules' if hoisted package
    const packageName = key.slice(key.lastIndexOf('node_modules/') + 13);
    mappedPackages[packageName] = mappedPackages[packageName] || {};

    const packageMeta = {
      path: key,
      dev,
      optional,
    };

    const newKey = packageName + '@' + value.version;
    if (mappedPackages[packageName][newKey]) {
      mappedPackages[packageName][newKey].packageMeta.push(packageMeta);
    } else {
      mappedPackages[packageName][newKey] = {
        ...value,
        packageMeta: [packageMeta],
      };
    }
  });
  return mappedPackages;
}

/**
 * Generates package-lock.json file from `LockFileData` object
 *
 * @param lockFile
 * @returns
 */
export function stringifyNpmLockFile(lockFileData: LockFileData): string {
  const dependencies = {};
  const packages: Dependencies = {
    '': lockFileData.lockFileMetadata.rootPackage,
  };

  const keys = Object.keys(lockFileData.dependencies);
  for (let i = 0; i < keys.length; i++) {
    const packageName = keys[i];
    const packageVersions = lockFileData.dependencies[packageName];
    const values = Object.values(packageVersions);

    values.forEach((value) => {
      unmapPackage(packages, value);
      unmapPackageDependencies(dependencies, packageName, value);
    });
  }

  // generate package lock JSON
  const lockFileJson: NpmLockFile = {
    ...lockFileData.lockFileMetadata.metadata,
    packages: sortObject(packages),
    dependencies: sortDependencies(dependencies),
  };

  return JSON.stringify(lockFileJson, null, 2) + '\n';
}

// remapping the package back to package-lock format
function unmapPackage(
  packages: Dependencies,
  { packageMeta, ...value }: PackageDependency
) {
  // we need to decompose value, to achieve particular field ordering
  const {
    version,
    resolved,
    integrity,
    license,
    devOptional,
    hasInstallScript,
    ...rest
  } = value;

  for (let i = 0; i < packageMeta.length; i++) {
    const { path, dev, optional } = packageMeta[i];
    // we are sorting the properties to get as close as possible to the original package-lock.json
    packages[path] = {
      version,
      resolved,
      integrity,
      dev,
      devOptional,
      hasInstallScript,
      license,
      optional,
      ...rest,
    };
  }
}

function unmapPackageDependencies(
  dependencies: Record<string, NpmDependency>,
  packageName: string,
  { packageMeta, ...value }: PackageDependency & { packageMeta: PackageMeta[] }
): void {
  const { version, resolved, integrity, devOptional } = value;

  for (let i = 0; i < packageMeta.length; i++) {
    const { path, dev, optional } = packageMeta[i];
    const projectPath = path.split('node_modules/').slice(1);

    const innerDeps = getProjectNodeAndEnsureParentHierarchy(
      projectPath,
      dependencies
    );
    const requires = unmapDependencyRequires(value);

    // sorting fields to match package-lock structure
    innerDeps[packageName] = {
      version,
      resolved,
      integrity,
      dev,
      devOptional,
      optional,
      requires,
      ...innerDeps[packageName],
    };
  }
}

// generates/ensures entire parent hierarchy exists for the given project path
// returns pointer to last project in the path
function getProjectNodeAndEnsureParentHierarchy(
  projects: string[],
  dependencies: Record<string, NpmDependency>
) {
  while (projects.length > 1) {
    const parentName = projects.shift().replace(/\/$/, '');
    if (!dependencies[parentName]) {
      dependencies[parentName] = {} as NpmDependency;
    }
    if (!dependencies[parentName].dependencies) {
      dependencies[parentName].dependencies = {};
    }
    dependencies = dependencies[parentName].dependencies;
  }
  return dependencies;
}

// combine dependencies and optionalDependencies into requires and sort them
function unmapDependencyRequires(
  value: Omit<PackageDependency, 'packageMeta'>
): Record<string, string> {
  if (!value.dependencies && !value.optionalDependencies) {
    return undefined;
  }
  const dependencies = {
    ...(value.dependencies || {}),
    ...(value.optionalDependencies || {}),
  };
  const sortedKeys = Object.keys(dependencies).sort();
  const result = {};
  for (let i = 0; i < sortedKeys.length; i++) {
    const key = sortedKeys[i];
    result[key] = dependencies[key];
  }
  return result;
}

// recursively sort dependencies
function sortDependencies(
  unsortedDependencies: Record<string, NpmDependency>
): Record<string, NpmDependency> {
  const dependencies = {};
  const sortedKeys = Object.keys(unsortedDependencies).sort();

  for (let i = 0; i < sortedKeys.length; i++) {
    const value = unsortedDependencies[sortedKeys[i]];
    dependencies[sortedKeys[i]] = value;
    if (value.dependencies) {
      value.dependencies = sortDependencies(value.dependencies);
    }
  }

  return dependencies;
}

/**
 * Prunes the lock file data based on the list of packages and their transitive dependencies
 *
 * @param lockFileData
 * @returns
 */
export function pruneNpmLockFile(
  lockFileData: LockFileData,
  packages: string[]
): LockFileData {
  // todo(meeroslav): This functionality has not been implemented yet
  console.warn(
    'Pruning package-lock.json is not yet implemented. Returning entire lock file'
  );
  return lockFileData;
}
