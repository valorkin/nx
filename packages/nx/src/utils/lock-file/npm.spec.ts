import { parseNpmLockFile, stringifyNpmLockFile } from './npm';
import { lockFile } from './__fixtures__/npm.lock';

describe('npm LockFile utility', () => {
  const parsedLockFile = parseNpmLockFile(lockFile);

  it('should parse lockfile correctly', () => {
    expect(parsedLockFile.lockFileMetadata).toEqual({
      metadata: {
        lockfileVersion: 2,
        name: 'test',
        requires: true,
        version: '0.0.0',
      },
      rootPackage: {
        devDependencies: {
          '@nrwl/cli': '14.7.5',
          '@nrwl/workspace': '14.7.5',
          nx: '14.7.5',
          prettier: '^2.6.2',
          typescript: '~4.8.2',
        },
        license: 'MIT',
        name: 'test',
        version: '0.0.0',
      },
    });
    expect(Object.keys(parsedLockFile.dependencies).length).toEqual(324);
    expect(
      parsedLockFile.dependencies['@ampproject/remapping']
    ).toMatchSnapshot();
    expect(parsedLockFile.dependencies['typescript']).toMatchSnapshot();
  });

  it('should map various versions of packages', () => {
    expect(
      Object.keys(parsedLockFile.dependencies['@jridgewell/gen-mapping']).length
    ).toEqual(2);
    expect(
      parsedLockFile.dependencies['@jridgewell/gen-mapping'][
        '@jridgewell/gen-mapping@0.1.1'
      ]
    ).toBeDefined();
    expect(
      parsedLockFile.dependencies['@jridgewell/gen-mapping'][
        '@jridgewell/gen-mapping@0.3.2'
      ]
    ).toBeDefined();
  });

  it('should map various instances of the same version', () => {
    const jestResolveDependency =
      parsedLockFile.dependencies['jest-resolve']['jest-resolve@28.1.3'];
    expect(jestResolveDependency.packageMeta.length).toEqual(2);
    expect((jestResolveDependency.packageMeta[0] as any).path).toEqual(
      'node_modules/jest-runner/node_modules/jest-resolve'
    );
    expect((jestResolveDependency.packageMeta[1] as any).path).toEqual(
      'node_modules/jest-runtime/node_modules/jest-resolve'
    );
  });

  it('should map optional field', () => {
    const tsDependency =
      parsedLockFile.dependencies['typescript']['typescript@4.8.3'];
    expect((tsDependency.packageMeta[0] as any).optional).toBeFalsy();
    const fsEventsDependency =
      parsedLockFile.dependencies['fsevents']['fsevents@2.3.2'];
    expect((fsEventsDependency.packageMeta[0] as any).optional).toBeTruthy();
  });

  it('should match the original file on stringification', () => {
    expect(stringifyNpmLockFile(parsedLockFile)).toEqual(lockFile);
  });
});
