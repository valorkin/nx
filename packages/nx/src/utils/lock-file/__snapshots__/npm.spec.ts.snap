// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`npm LockFile utility should parse lockfile correctly 1`] = `
Object {
  "@ampproject/remapping@2.2.0": Object {
    "dependencies": Object {
      "@jridgewell/gen-mapping": "^0.1.0",
      "@jridgewell/trace-mapping": "^0.3.9",
    },
    "engines": Object {
      "node": ">=6.0.0",
    },
    "integrity": "sha512-qRmjj8nj9qmLTQXXmaR1cck3UXSRMPrbsLJAasZpF+t3riI71BXed5ebIOYwQntykeZuhjsdweEc9BxH5Jc26w==",
    "packageMeta": Array [
      Object {
        "dev": true,
        "optional": undefined,
        "path": "node_modules/@ampproject/remapping",
      },
    ],
    "resolved": "https://registry.npmjs.org/@ampproject/remapping/-/remapping-2.2.0.tgz",
    "version": "2.2.0",
  },
}
`;

exports[`npm LockFile utility should parse lockfile correctly 2`] = `
Object {
  "typescript@4.8.3": Object {
    "bin": Object {
      "tsc": "bin/tsc",
      "tsserver": "bin/tsserver",
    },
    "engines": Object {
      "node": ">=4.2.0",
    },
    "integrity": "sha512-goMHfm00nWPa8UvR/CPSvykqf6dVV8x/dp0c5mFTMTIu0u0FlGWRioyy7Nn0PGAdHxpJZnuO/ut+PpQ8UiHAig==",
    "packageMeta": Array [
      Object {
        "dev": true,
        "optional": undefined,
        "path": "node_modules/typescript",
      },
    ],
    "resolved": "https://registry.npmjs.org/typescript/-/typescript-4.8.3.tgz",
    "version": "4.8.3",
  },
}
`;
